google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/RateLimitingServerStreamingCallable.java

```diff
@@ -0,0 +1,149 @@
+ */
+package com.google.cloud.bigtable.data.v2.stub;
+
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.DeadlineExceededException;
+import com.google.api.gax.rpc.ResourceExhaustedException;
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.ServerStreamingCallable;
+import com.google.api.gax.rpc.StreamController;
+import com.google.api.gax.rpc.UnavailableException;
+import com.google.bigtable.v2.MutateRowsRequest;
+import com.google.bigtable.v2.MutateRowsResponse;
+import com.google.bigtable.v2.RateLimitInfo;
+import com.google.cloud.bigtable.data.v2.stub.metrics.BigtableTracer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import com.google.common.util.concurrent.RateLimiter;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+import org.threeten.bp.Duration;
+
+class RateLimitingServerStreamingCallable
+    extends ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> {
+  private static final Logger logger =
+      Logger.getLogger(RateLimitingServerStreamingCallable.class.getName());
+  private static final long DEFAULT_QPS = 10;
+  // Default interval before changing the QPS on error response
+  private static Duration DEFAULT_PERIOD = Duration.ofSeconds(10);
+  private final RateLimiter limiter;
+  private final AtomicLong lastQpsChangeTime = new AtomicLong(System.currentTimeMillis());
+  private final ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> innerCallable;
+  private static final double MIN_QPS = 0.1;
+  private static final double MAX_QPS = 100_000;
+  @VisibleForTesting static final double MIN_FACTOR = 0.7;
+  private static final double MAX_FACTOR = 1.3;
+
+  RateLimitingServerStreamingCallable(
+      @Nonnull ServerStreamingCallable<MutateRowsRequest, MutateRowsResponse> innerCallable) {
+    this.limiter = RateLimiter.create(DEFAULT_QPS);
+    this.innerCallable = Preconditions.checkNotNull(innerCallable, "Inner callable must be set");
+    logger.info("Rate limiting is enabled with initial QPS of " + limiter.getRate());
+  }
+
+  @Override
+  public void call(
+      MutateRowsRequest request,
+      ResponseObserver<MutateRowsResponse> responseObserver,
+      ApiCallContext context) {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    limiter.acquire();
+    stopwatch.stop();
+    if (context.getTracer() instanceof BigtableTracer) {
+      ((BigtableTracer) context.getTracer())
+          .batchRequestThrottled(stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+    RateLimitingResponseObserver innerObserver = new RateLimitingResponseObserver(responseObserver);
+    innerCallable.call(request, innerObserver, context);
+  }
+
+  class RateLimitingResponseObserver extends SafeResponseObserver<MutateRowsResponse> {
```

Comment:
static?

Reply:
Can we defer this ? :( the experiments are run without static :( and I worry making it global may change some results